### PR TITLE
Implement StaticSVI

### PR DESCRIPTION
**Pull request by fehiepsi**
This pull request implements StaticSVI, which is an SVI interface for model & guide which does not create new params dynamically. Hence LBFGS works with this inference (address #1519).

### CODE REVIEW

File: tests/infer/test_static_svi.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+            pyro.sample("obs", dist.Bernoulli(p_latent), obs=data)
+        return p_latent
+
+    def guide():
+        alpha_q_log = pyro.param("alpha_q_log", log_alpha_n + 0.17)
+        beta_q_log = pyro.param("beta_q_log", log_beta_n - 0.143)
+        alpha_q, beta_q = torch.exp(alpha_q_log), torch.exp(beta_q_log)
+        pyro.sample("p_latent", dist.Beta(alpha_q, beta_q))
+
+    adam = optim.Adam({"lr": .001})
```
Comment (fritzo) [inline tests/infer/test_static_svi.py]: Can you also add a test using `LBFGS`?

**Comment by eb8680** [comment]
I don't really think this is necessary.  In this situation, we should encourage users to write their models as `nn.Module`s and use Pyro losses and `torch` optimizers together directly instead of going through the SVI and `PyroOptim` interfaces:
```python
class MyModel(nn.Module):
    def __init__(self, ...):
        ...

    def model(self, batch):
        ....

    def guide(self, batch):
        ...

model = MyModel(...)

elbo = pyro.infer.Trace_ELBO()
optim = torch.optim.SGD(model.parameters())  # or LBFGS etc
for batch in data:
    optim.zero_grad()
    elbo.loss_and_grads(model.model, model.guide, batch)
    optim.step()

**Comment by fehiepsi** [comment]
@eb8680 Do you mean that we'll have a Pyro `nn.Module` which will catch all parameters from `model` and `guide` when we call `mymodel.parameters()`? Without that, users have to know about `poutine.trace(param_only=True)` to capture all params for pytorch optimizer. In addition, `SVI.run()` is quite convenient to get samples from posterior. Otherwise, users will have to learn about `poutine.trace(...)`, `poutine.replay(...)`,...

File: tests/infer/test_static_svi.py (hunk 1 above)
Comment (fehiepsi) [reply tests/infer/test_static_svi.py]: @fritzo I have added another test for it. Using LBFGS for this model is quite flaky.

**Comment by eb8680** [comment]
@fehiepsi `module.parameters()` is a method of `torch.nn.Module` that PyTorch users would be familiar with.  Most of our tutorials, e.g. the VAE tutorial, already wrap the model and guide as methods of a single `nn.Module` and never call `pyro.param` explicitly in the model or guide, but they use `pyro.module`  (which calls `module.named_parameters()` under the hood) to pass the parameters to the Pyro parameter store.  I don't think that extra layer of indirection is buying us much, and I think the snippet above is less opaque and more consistent with PyTorch idioms.

We shouldn't try to fix leaky abstractions by introducing more indirection, we should just deprecate/remove them.  `SVI`, `Trace`, `ParamStoreDict`, and `PyroOptim` are some of the worst offenders.  On that note, re: `trace` and `replay` for parameter capture and posterior sampling instead of `SVI`, I actually think it'd be better to do that.  We can always provide a couple of tiny wrappers that remove boilerplate without obfuscation.

**Comment by fehiepsi** [comment]
@eb8680 I took me a while to have a feeling that I understand what you mean. :) Did you mean that we define all the parameters ahead, so we don't need to use ParamStoreDict, PyroOptim? If that is the case, then I am in with the future deprecation of these wrappers (together with SVI of course). Please correct me if my understanding is incorrect.

About this PR, how about moving this to contrib for a while?

**Comment by fehiepsi** [comment]
For that purpose, I vote to close this PR. I'll use that pattern in a GP tutorial instead. :)

**Pull request closed without merging**